src/ILCompiler.Compiler/src/Compiler/MetadataGeneration.cs

```diff
@@ -109,14 +113,7 @@ private void Graph_NewMarkedNode(DependencyNodeCore<NodeFactory> obj)
             {
                 _typesWithEETypesGenerated.Add(eetypeNode.Type);
                 AddGeneratedType(eetypeNode.Type);
-
-                // If this is an instantiated non-canonical generic type, add it to the generic instantiations hashtable
-                if (eetypeNode.Type.HasInstantiation && !eetypeNode.Type.IsGenericDefinition)
-                {
-                    if (!eetypeNode.Type.IsCanonicalSubtype(CanonicalFormKind.Any))
-                        GenericsHashtable.AddInstantiatedTypeEntry(_nodeFactory, eetypeNode.Type);
-                }
-
+                GenericsHashtable.AddInstantiatedTypeEntry(_nodeFactory, eetypeNode.Type);
```

Comment:
The reason i did it this way is because i personally like the idea of an API that knows how to handle its inputs, rather than having the caller worry about the right conditions to call it (also useful if the api will be called from different places). That said, looks like the name of the api is not very clear, and should change to something like AddEntryIfEligible.